src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp

```diff
@@ -25,17 +25,404 @@ namespace XBU = XBUtilities;
+  if (!currentDSA.name.empty()) {
+    same_dsa = ((candidate.name == currentDSA.name) &&
+      (candidate.matchId(currentDSA)));
+    same_bmc = ((currentDSA.bmcVer.empty()) ||
+      (candidate.bmcVer == currentDSA.bmcVer));
+  }
+  if (same_dsa && same_bmc) {
+    return DSAInfo("");
+  }
+
+  // Header info
+  // May need to refactor this when adding functionality for multiple devices
+  std::cout << boost::format("%s : %d\n") % "Device BDF" % flasher.sGetDBDF();
+  std::cout << "Current Configuration\n";
+
+  std::cout << boost::format("  %-20s : %s\n") % "Platform" % currentDSA.name;
+  std::cout << boost::format("  %-20s : %s\n") % "SC Version" % currentDSA.bmcVer;
+  std::cout << boost::format("  %-20s : 0x%x\n") % "Platform ID" % currentDSA.timestamp;
+
+  std::cout << "\nIncoming Configuration\n";
+  std::pair <std::string, std::string> s = deployment_path_and_filename(candidate.file);
+  std::cout << boost::format("  %-20s : %s\n") % "Deployment File" % s.first;
+  std::cout << boost::format("  %-20s : %s\n") % "Deployment Directory" % s.second;
+  std::cout << boost::format("  %-20s : %s\n") % "Size" % file_size(candidate.file.c_str());
+  std::cout << boost::format("  %-20s : %s\n\n") % "Timestamp" % get_file_timestamp(candidate.file.c_str());
+
+  std::cout << boost::format("  %-20s : %s\n") % "Platform" % candidate.name;
+  std::cout << boost::format("  %-20s : %s\n") % "SC Version" % candidate.bmcVer;
+  std::cout << boost::format("  %-20s : 0x%x\n\n") % "Platform ID" % candidate.timestamp;
+
+  std::cout << "----------------------------------------------------\n";
+  std::cout << "Actions to perform:\n";
+  if(!same_dsa)
+    std::cout << "  -Program flash image\n";
+  if(!same_bmc)
+    std::cout << "  -Program SC image\n";
+  std::cout << "----------------------------------------------------\n";
+  // end header info
+
+  return candidate;
+}
+
+/* 
+ * Confirm with the user
+ * Helper method for auto_flash
+ */
+bool canProceed()
+{
+  std::string input;
+  bool answered = false;
+  bool proceed = false;
+
+  while (!answered) {
+    std::cout << "Are you sure you wish to proceed? [y/n]: ";
+    std::cin >> input;
+    if(input.compare("y") == 0 || input.compare("n") == 0)
+      answered = true;
+  }
+
+  proceed = (input.compare("y") == 0);
+  if (!proceed)
+    std::cout << "Action canceled." << std::endl;
+  return proceed;
+}
+
+/* 
+ * Flash shell and sc firmware
+ * Helper method for auto_flash
+ */
+static int 
+updateShellAndSC(uint16_t boardIdx, DSAInfo& candidate, bool& reboot)
+{
+  reboot = false;
+
+  Flasher flasher(boardIdx);
+  if(!flasher.isValid())
+    throw xrt_core::error(boost::str(boost::format("%d is an invalid index") % boardIdx));
+
+  bool same_dsa = false;
+  bool same_bmc = false;
+  DSAInfo current = flasher.getOnBoardDSA();
+  if (!current.name.empty()) {
+    same_dsa = (candidate.name == current.name &&
+      candidate.matchId(current));
+    same_bmc = (current.bmcVer.empty() ||
```

Comment:
That's because we don't want to flash when bmc is unknown. David fixed this in xbmgmt in https://github.com/Xilinx/XRT/commit/77345c185a7a642a96c491828640e23f64051af3. But yes, this code is wrong, I'll take it out.